### PR TITLE
fix #255: Add new rule require-v-if-else-key

### LINKED ISSUE #255: Rule proposal: `v-if-else-key`
**Style guide:**
https://vuejs.org/v2/style-guide/#v-if-v-if-else-v-else-without-key-use-with-caution

**Description:**
This rule would enforce to add `key` property on elements with `v-if` and `v-else` directives if they're both of the same type.

### PR DESCRIPTION
**Pull request by doug-wade**
This one's probably a smidge later than it could have been -- I can't find it in the style guide anymore -- but better late than never?

### CODE REVIEW

File: lib/rules/require-v-if-else-key.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+      if (!keyDirective) {
+        return ''
+      }
+
+      if (
+        !keyDirective.value ||
+        !keyDirective.value.expression ||
+        keyDirective.value.expression.type !== 'Identifier'
+      ) {
+        return null
```
Comment (doug-wade) [inline lib/rules/require-v-if-else-key.js:61]: I'm not sure about using the empty string and null to represent two different things. I'm definitely open to other suggestions.

File: docs/rules/require-v-if-else-key.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+</eslint-code-block>
+
+## :wrench: Options
+
+Nothing.
+
+## :books: Further Reading
+
+- [Guide (for v2) - v-if without key](https://v2.vuejs.org/v2/style-guide/?redirect=true#v-if-v-else-if-v-else-without-key-use-with-caution)
```
Comment (FloEdelmann) [inline docs/rules/require-v-if-else-key.md:47]:
```suggestion
- [Guide (for v2) - v-if without key](https://v2.vuejs.org/v2/style-guide/#v-if-v-else-if-v-else-without-key-use-with-caution)
```

File: lib/rules/require-v-if-else-key.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+module.exports = {
+  meta: {
+    type: 'problem',
+    docs: {
+      description: 'require `key` with `v-if/v-else-if/v-else` directives',
```
Comment (FloEdelmann) [inline lib/rules/require-v-if-else-key.js:27]:
```suggestion
      description: 'require `key` with `v-if`/`v-else-if`/`v-else` directives',
```

File: lib/rules/require-v-if-else-key.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+// ------------------------------------------------------------------------------
+
+const utils = require('../utils')
+
+// ------------------------------------------------------------------------------
+// Helpers
+// ------------------------------------------------------------------------------
+
+// ...
+
```
Comment (FloEdelmann) [inline lib/rules/require-v-if-else-key.js:18]: Please remove these empty boilerplate lines.


```suggestion
```

File: docs/rules/require-v-if-else-key.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+---
+# vue/require-v-if-else-key
+
+> require `key` with `v-if/v-else-if/v-else` directives
+
+- :exclamation: <badge text="This rule has not been released yet." vertical="middle" type="error"> ***This rule has not been released yet.*** </badge>
+
+## :book: Rule Details
+
+This rule reports the elements which have `v-if`, `v-else-if`, and/or `v-else` and do not have a `v-bind:key` or `key`.
```
Comment (FloEdelmann) [inline docs/rules/require-v-if-else-key.md:15]: An element can't have more than one of these directives. Also, it should be clarified that they have to have the same tag name.

```suggestion
This rule reports consecutive elements with the same tag name which all have `v-if`, `v-else-if`, or `v-else` and do not have a `v-bind:key` or `key`.
```

File: lib/rules/require-v-if-else-key.js [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    docs: {
+      description: 'require `key` with `v-if/v-else-if/v-else` directives',
+      categories: undefined,
+      url: 'https://eslint.vuejs.org/rules/require-v-if-else-key.html'
+    },
+    fixable: null,
+    schema: [],
+    messages: {
+      unexpected:
+        'Elements in v-if/v-else-if/v-else expect to have distinct keys if they are of the same type.'
```
Comment (FloEdelmann) [inline lib/rules/require-v-if-else-key.js:35]: Maybe _tag name_ is a bit clearer than _element type_?


```suggestion
        'Elements in v-if/v-else-if/v-else should have distinct keys if they have the same tag name.'
```

File: lib/rules/require-v-if-else-key.js [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+  create(context) {
+    /**
+     * Gets the key from an element.
+     * @param {VElement} element The element node to get the key from.
+     * @return {String | null}
+     */
+    function getKey(element) {
+      const keyAttribute = utils.getAttribute(element, 'key')
+      if (keyAttribute) {
+        return keyAttribute.value ? keyAttribute.value.value || '' : ''
```
Comment (FloEdelmann) [inline lib/rules/require-v-if-else-key.js:48]:
```suggestion
        return keyAttribute.value?.value || ''
```

File: tests/lib/rules/require-v-if-else-key.js [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+      filename: 'test.vue',
+      code: `
+      <template>
+        <div>
+          <div v-if="conditional" key="foo">foo</div>
+          <div v-else-if="conditional" key="bar">bar</span>
+        </div>
+      </template>
+      `
+    }
```
Comment (FloEdelmann) [inline tests/lib/rules/require-v-if-else-key.js:84]: Please add one valid test case with two independent `<div v-if="...">`s.

**Review by FloEdelmann — CHANGES_REQUESTED** [review]
Thanks for this rule. I have a few suggestions.

File: tests/lib/rules/require-v-if-else-key.js (hunk 8 above)
Comment (ota-meshi) [inline tests/lib/rules/require-v-if-else-key.js:84]: We need to add tests using `<template v-else>`. Note that vue2 does not allow you to place a key in `<template>`.

File: lib/rules/require-v-if-else-key.js [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        return
+      }
+
+      const key = getKey(element)
+
+      if (key === null) {
+        return
+      }
+
+      if (!key || key === getKey(prevSibling)) {
```
Comment (ota-meshi) [inline lib/rules/require-v-if-else-key.js:88]: I don't think it is necessary to check the contents of the key. If the user uses the same key, I think the user is intentional.
So, I think we can change it to check if the element has a key.

File: docs/rules/require-v-if-else-key.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
```
Comment (ota-meshi) [inline docs/rules/require-v-if-else-key.md:16]: Can you explain here that Vue3 doesn't require this rule? In Vue3, the key is automatically assigned.

**Review by ota-meshi — CHANGES_REQUESTED** [review]

File: docs/rules/require-v-if-else-key.md (hunk 10 above)
Comment (ota-meshi) [reply docs/rules/require-v-if-else-key.md:16]: Demo: https://vue-next-template-explorer.netlify.app/#eyJzcmMiOiI8ZGl2IHYtaWY9XCJlcnJvclwiPlxuICBFcnJvcjoge3sgZXJyb3IgfX1cbjwvZGl2PlxuPGRpdiB2LWVsc2U+XG4gIHt7IHJlc3VsdHMgfX1cbjwvZGl2PlxuIiwic3NyIjpmYWxzZSwib3B0aW9ucyI6eyJvcHRpbWl6ZUltcG9ydHMiOmZhbHNlLCJvcHRpbWl6ZUJpbmRpbmdzIjpmYWxzZX19

**Review by FloEdelmann — APPROVED** [review]
Please run `npm run update` to uodate the docs and fix lint any potential lint issues.

Other than that, this looks good to me. Thanks!

File: tests/lib/rules/require-v-if-else-key.js [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+            <div>foo</div>
+            <div>bar</span>
+          </template>
+          <template v-else>
+            <div>baz</div>
+            <div>quux</span>
+          </template>
+        </div>
+      </template>
+      `
```
Comment (ota-meshi) [inline tests/lib/rules/require-v-if-else-key.js:142]: Why this is valid? I think this should be reported.

Comment (doug-wade) [reply tests/lib/rules/require-v-if-else-key.js:142]: I must have misinterpreted [your comment](https://github.com/vuejs/eslint-plugin-vue/pull/1845#discussion_r850099546):

> Note that vue2 does not allow you to place a key in <template>.

If no keys are allowed, mustn't we allow templates without keys?

Comment (ota-meshi) [reply tests/lib/rules/require-v-if-else-key.js:142]: In Vue2 need to put the key on the actual element, not the `<template>`.

e.g.

```vue
<template v-if="foo">
  <div key="a" />
</template>
<template v-else>
  <div key="b" />
</template>
```

```vue
<template v-if="foo">
  <div key="a" />
  <div key="b" />
</template>
<template v-else>
  <div key="c" />
  <div key="d" />
</template>
```

However, with Vue3 you can do the following:



```vue
<template v-if="foo" :key="a">
  <div />
</template>
<template v-else :key="b">
  <div />
</template>
```

```vue
<template v-if="foo" :key="a">
  <div />
  <div />
</template>
<template v-else :key="b">
  <div />
  <div />
</template>
```

**Comment by doug-wade** [comment]
Heya, things have gotten pretty busy for me, and I don't think I'll have time to clean this one up in the near future. I'm going to close this one for now.

**Pull request closed without merging**